### DISCUSSION TITLE
How to upload firmware to a system without a SDCard

### DISCUSSION THREAD

**Discussion by nkokkos** [nerves-project/nerves#869] [General]
Hi.

First of all, congratulations on your work on nerves. This is a wonderful system.

I am just learning nerves and was wondering if there is way to burn the firmware on the flash of an embedded system. That is, from the boards supported by nerves, I see that most of them have a sdcard slot and nerves is used to burn the firmware on the card and then the system boots from the card.

What about if there is, say, a system you provide to your customers and you don't want them to fiddle with the sdcard. A system that has 
enough flash memory to run nerves. How do you flash the memory of that system with nerves? Is there an example of how you can do  this? Perhaps, I have not read all the documentation on nerves, but I was wondering if you could provide some pointers.

Thank you so much in advance.

**Comment by fhunleth** [comment]
Hi @nkokkos,

We try to officially support hardware that's easy for people to buy. Those tend to have MicroSD cards on them as you saw. Many companies using Nerves and creating their own hardware choose to use eMMC parts instead of MicroSD cards. It really depends on the processor that they use for how they load the images. It can be over USB, through some debug connector or they can have a MicroSD slot mounted for development use. The ones that I'm most familiar with are Allwinner and TI AM335x (like the BeagleBone) custom boards. The Allwinner ones can be put into USB FEL mode and let you program an eMMC part (or do many other things) via a USB cable. The BeagleBone Black design has both a MicroSD and eMMC, so if you copy that, it's possible to make a loader MicroSD card that copies the firmware over to eMMC.

What I'm describing is the first time load of firmware. After you get a good image on the board, you'd typically want to ssh your firmware updates over or update via some easier way. Nerves firmware updates are single files with the extension `.fw`. The update process is to copy that `.fw` file over somehow and the run a program called `fwup` to apply it and then to reboot. The Nerves tooling can automate this, but that's all that it is underneath the covers.

Hope this helps.

Also, we don't monitor GitHub discussions and I hadn't realized it was on. Sorry about that. That's why it took a bit to see and respond. We typically communicate via the Elixir Slack and the Elixir Forum.

**Reply by nkokkos** [reply]
@fhunleth 

Thank you so much for taking the time to reply. Your response was very analytical and helpful. Thank you again.
